service/firewall/packet_handler.go

```diff
@@ -581,6 +586,97 @@ func inspectAndVerdictHandler(conn *network.Connection, pkt packet.Packet) {
 	issueVerdict(conn, pkt, 0, true)
 }
 
+func inspectDNSPacket(conn *network.Connection, pkt packet.Packet) {
+	// Ignore info-only packets in this handler.
+	if pkt.InfoOnly() {
+		return
+	}
+
+	dnsPacket := new(dns.Msg)
+	err := pkt.LoadPacketData()
+	if err != nil {
+		_ = pkt.Block()
+		log.Errorf("filter: failed to load packet payload: %s", err)
+		return
+	}
+
+	// Parse and block invalid packets.
+	err = dnsPacket.Unpack(pkt.Payload())
+	if err != nil {
+		err = pkt.PermanentBlock()
+		if err != nil {
+			log.Errorf("filter: failed to block packet: %s", err)
+		}
+		_ = conn.SetVerdict(network.VerdictBlock, "none DNS data on DNS port", "", nil)
+		conn.VerdictPermanent = true
+		conn.Save()
+		return
+	}
+
+	// Packet was parsed, accept the connection and continue.
+	err = pkt.Accept()
+	if err != nil {
+		log.Errorf("filter: failed to accept dns packet: %s", err)
+	}
+
+	conn.Type = network.DNSRequest
```

Comment:
The connection technically is still an IP connection. This should not be changed.
The DNSRequest ist only used for actual DNS requests that came in through the nameserver.

As also, the connection can carry multiple DNS requests.